spices/SPICE-0009-external-readers.adoc

```diff
@@ -0,0 +1,319 @@
+email = "appleseed@example.com"
+----
+
+For security reasons, schemes provided by external readers are not allowed by default and clients must explicitly allow the desired scheme(s).
+In this example, the external reader may provide both `ldap:` and `ldaps:` schemes.
+If a codebase or user wishes to enforce that only secure LDAP connections are made, the `ldaps:` scheme may be allowed while `ldap:` is not.
+
+== Detailed design
+
+To avoid terminology confusion with the existing language binding message passing model, the `pkl` process will remain known as the "server" and the reader process will be known as the "client", despite all requests in this proposal being initiated by the `pkl` process.
+
+=== Overall Flow
+
+1. User requests an evaluator (via Java API, Message Passing API, or CLI) with an external reader.
+2. Entrypoint (client Java code, `CliEvaluator`, or `Server`) instantiates an `ExternalReader`.
+3. The `ExternalReader` is started, which starts the configured child process and sends a `DiscoverReadersRequest` message.
+4. A `DiscoverReadersResponse` message is received (within some timeout).
+5. The entrypoint requests resource readers (`ExternalResourceReader`) and module key factories (`ExternalModuleKeyFactory`) from the `ExternalReader` and uses them to build an evaluator.
+`ExternalResourceReader` and `ExternalModuleKeyFactory` function similarly to the existing `ClientResourceReader` and `ClientModuleKeyFactory`, implementation may be reusable entirely.
+6. Evaluation begins.
+7. A resource or module read is encountered.
+8. The appropriate `Read*Request` message is sent to the child process, the corresponding `Read*Response` message is awaited.
+9. Evaluation continues.
+10. Evaluation completes.
+11. The `ExternalReader` is closed, terminating the child process.
+
+=== Java API
+
+New APIs:
+
+* `ExternalResourceReader` - a `ResourceReader` implementation similar (identical?) to `ClientResourceReader`.
+* `ExternalModuleKeyFactory` - a `ModuleKeyFactory` implementation similar (identical?) to `ClientModuleKeyFactory`.
+* `ExternalModuleKey` - a `ModuleKey` implementation similar (identical?) to `ClientModuleKey`.
+* `ExternalReader` - manages the lifecycle of child processes.
+    ** Explicit `open`/`close` methods to manage child process lifecycle.
+    ** Opening an `ExternalReader` spawns the subprocess, sets up the `MessageTransport`, sends a `DiscoverReadersRequest` message, and awaits a `DiscoverReadersResponse` response.
+    ** Methods to procure `ExternalResourceReader` and `ExternalModuleKeyFactory` instances.
+
+This proposal requires that the message passing API functionality move out of pkl-server.
+A new project pkl-msgapi will be created to contain the new APIs and the core messaging code current part of pkl-server (`pkl-server/src.main/kotlin/org.pkl.server/Message*.kt`).
+The pkl-msgapi project will be depended on by pkl-cli, pkl-server, and Java API clients using external readers.
```

Comment:
Might be better to just move the message passing API into pkl-core altogether.

Reply:
Sounds good, updated! I also realized that allowing configuration of external readers via PklProject would most likely necessitate these being in pkl-core.